### ISSUE TITLE
[Dynamo] Graph Re-compilation Invoked by `np.sqrt`

### ISSUE THREAD

**Issue by alpha0422** [pytorch/pytorch#91101]
### 🐛 Describe the bug

We find that `np.sqrt` can lead to graph re-compilation, but `math.sqrt` will not. I have brief search in the Github issues but didn't find realted issue, so I think this might be one bug.

Here's one minimal example to reproduce:

```python
#!/usr/bin/env python

import sys
import torch
from torch._dynamo import optimize, config
config.cache_size_limit = 4
torch.manual_seed(0)

class Op(torch.autograd.Function):
    @staticmethod
    def forward(ctx, i):
        result = i * sqrt(i.numel())
        ctx.save_for_backward(result)
        return result

    @staticmethod
    def backward(ctx, grad_output):
        result, = ctx.saved_tensors
        return grad_output * sqrt(result.numel())

@optimize("inductor")
def toy_example(x):
    return Op.apply(x).norm()

def test():
    for i in range(5):
        x = torch.randn(10)
        print(toy_example(x))

if __name__ == "__main__":
    msg = "Usage: python bug-2.py numpy|math"
    assert len(sys.argv) == 2, msg
    if sys.argv[1] == "numpy":
        from numpy import sqrt
    elif sys.argv[1] == "math":
        from math import sqrt
    else:
        raise  Exception(msg)
    test()
```

Here's the message I got with PyTorch `1.14.0a0+410ce96`:

```bash
$ python bug-2.py numpy
tensor(11.0491)
tensor(8.3844)
tensor(9.0465)
tensor(9.7683)
[2022-12-19 14:51:04,617] torch._dynamo.convert_frame: [WARNING] torch._dynamo hit config.cache_size_limit (4)
   function: '<graph break in forward>' (bug-2.py:12)
   reasons:  ['___guarded_code.valid']
to diagnose recompilation issues, see https://github.com/pytorch/torchdynamo/blob/main/TROUBLESHOOTING.md.
tensor(8.0597)

$ python bug-2.py math
tensor(11.0491)
tensor(8.3844)
tensor(9.0465)
tensor(9.7683)
tensor(8.0597)
```

### Versions

```
Collecting environment information...
PyTorch version: 1.14.0a0+410ce96
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-99-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-80GB
Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] open-clip-torch==2.7.0
[pip3] pytorch-quantization==2.1.2
[pip3] torch==1.14.0a0+410ce96
[pip3] torch-tensorrt==1.3.0a0
[pip3] torchtext==0.13.0a0+fae8e8c
[pip3] torchvision==0.15.0a0
[conda] Could not collect
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jjsjann123 @kevinstephano

**Comment by yanboliang** [comment]
@alpha0422 This is a bug, but not caused by ```np.sqrt```. ```torch._dynamo``` support python builtin function ```math.sqrt``` already, but doesn't support ```np.sqrt``` yet. So it causes a graph break and fall back to the python implementation. This should works well, no re-compilation.
The root cause is ```torch._dynamo``` doesn't support there is graph break inside of ```torch.autograd.Function.forward```. As after the graph break, the resume execution has guard on ```ctx.save_for_backward``` function(see below), who keeps failing on the next iteration as ```ctx``` was identified as ```UserDefinedObjectVariable```by mistake. I sent #91295 and verified it can fix this bug.
```
 local 'ctx.save_for_backward' FUNCTION_MATCH
  {
      'guard_types': ['ID_MATCH'],
      'code': ['___check_obj_id(ctx.save_for_backward, 140114533537280)'],
      'obj_weakref': None
      'guarded_class': <weakref at 0x7f6f96796310; to 'type' at 0x71fd40 (method)>
  }
```

**Comment by alpha0422** [comment]
@yanboliang, thanks for your fix and explanation on this, that helps. I applied your change and it works like a charm for this issue. However, it failed to pass another customized operator issue I filed [91102](https://github.com/pytorch/pytorch/issues/91102). The error messages without/with the fix are:

```console
$ python bug-3.py
tensor(3.7900)
tensor(6.5101)
tensor(10.7158)
tensor(22.0629)
[2023-01-11 08:48:53,067] torch._dynamo.convert_frame: [WARNING] torch._dynamo hit config.cache_size_limit (4)
   function: 'forward' (bug-3.py:9)
   reasons:  ['___guarded_code.valid']
to diagnose recompilation issues, see https://github.com/pytorch/torchdynamo/blob/main/TROUBLESHOOTING.md.
tensor(3.5327)

$ # apply the fix

$ python bug-3.py
Traceback (most recent call last):
  File "bug-3.py", line 34, in <module>
    test()
  File "bug-3.py", line 30, in test
    toy_example(x, dy)
  File "/usr/local/lib/python3.8/dist-packages/torch/_dynamo/eval_frame.py", line 173, in _fn
    return fn(*args, **kwargs)
  File "bug-3.py", line 22, in toy_example
    y = Exp.apply(x)
  File "bug-3.py", line 23, in <graph break in toy_example>
    y.backward(dy)
  File "/usr/local/lib/python3.8/dist-packages/torch/_tensor.py", line 450, in backward
    torch.autograd.backward(
  File "/usr/local/lib/python3.8/dist-packages/torch/autograd/__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/usr/local/lib/python3.8/dist-packages/torch/autograd/function.py", line 270, in apply
    return user_fn(self, *args)
  File "bug-3.py", line 17, in backward
    result, = ctx.saved_tensors
ValueError: not enough values to unpack (expected 1, got 0)
```

Could you take a look and resolve? Thanks!

**Comment by yanboliang** [comment]
@alpha0422 I hit the same issue as you posted here, that's why I still didn't land #91295. I'm prioritizing this work now and hope to fix it soon.

**Comment by yanboliang** [comment]
@alpha0422 #91295 has been merged, it fixed both this issue and #91102. Let me know if you have more questions.

**Comment by alpha0422** [comment]
Excellent! I tried and it fixes both issues, thanks for your help on this.
